hail/src/main/scala/is/hail/cxx/TranslationUnit.scala

```diff
@@ -0,0 +1,55 @@
+package is.hail.cxx
+
+import is.hail.nativecode._
+import is.hail.utils.ArrayBuilder
+
+class TranslationUnit(preamble: String, definitions: Array[Definition]) {
+
+  val functions: Map[String, Function] =
+    definitions
+      .filter(_.isInstanceOf[Function])
+      .map(f => (f.name, f.asInstanceOf[Function]))
+      .toMap
+
+  def source: String =
+    new PrettyCode(s"""$preamble
+       |
+       |NAMESPACE_HAIL_MODULE_BEGIN
+       |
+       |${definitions.map(d => if (d.isInstanceOf[Function]) d.define else s"${d.define};").mkString("\n\n")}
+       |
+       |NAMESPACE_HAIL_MODULE_END
+     """.stripMargin).toString()
+
+  def build(options: String): NativeModule = {
+    val st = new NativeStatus()
+    val mod = new NativeModule(options, source)
+    mod.findOrBuild(st)
+    assert(st.ok, st.toString())
+    mod
+  }
+}
+
+class TranslationUnitBuilder() {
+
+  val definitions: ArrayBuilder[Definition] = new ArrayBuilder[Definition]()
+
+  val includes: ArrayBuilder[String] = new ArrayBuilder[String]()
+
+  def include(header: String): Unit = {
```

Comment:
Option: `tub += Include("\"foo.h\"")`.  Or even `tub += SysInclude("vector")` and `tub += LocalInclude("foo.h")`.